usl_models/usl_models/flood_ml/model.py

```diff
@@ -118,12 +173,82 @@ def forward(
+
+        Args:
+            inputs: A dictionary of input tensors.
+                - "geospatial" (required): [H, W, f]
+                - "temporal" (required): [T]
+                - "spatiotemporal" (optional): a single flood map [H, W]
+
+        Returns:
+            A tuple (all_flood_predictions, max_flood_map) of the flood predictions and
+            the max predicted flood depth at each cell over the entire storm duration.
+            The maps are the same shape as the input: (H, W).
+        """
+        # TODO: Add data validation
+        geo_input = inputs.get("geospatial")
+        full_temp_input = data_utils.temporal_window_view(
+            inputs.get("temporal"), self.params["m_rainfall"]
+        )
+        # Note: We assume that, if provided, this input is a *single* flood map.
+        # For convenience, we add this to self.flood_maps to align indexing
+        # between flood maps and rainfall ,i.e., the current flooding conditions
+        # and rainfall at time t are stored at index t along the temporal axis.
+        st_input = inputs.get("spatiotemporal")
+        if st_input is None:
+            st_shape = geo_input.shape[:3] + [1]
+            st_input = tf.zeros(st_shape)
+        self.flood_maps.write(0, st_input)
+
+        # We use 1-indexing for simplicity. Time step t represents the t-th flood
+        # prediction.
+        for t in tf.range(1, self.n_predictions + 1):
+            st_input, temp_input = self._update_temporal_inputs(full_temp_input, t)
+            prediction = self.forward(st_input, geo_input, temp_input)
+            self.flood_maps.write(t, prediction)
+
+        # Concatenate the predictions.
+        # This gathers the predictions along axis 0, so we need to permute the
+        # time (0) and batch (1) axes.
+        predictions = self.flood_maps.gather(tf.range(1, self.n_predictions + 1))
+        self.flood_maps.close()
+        predictions = tf.transpose(predictions, perm=[1, 0, 2, 3, 4])
+        predictions = tf.squeeze(predictions, axis=-1)
+        max_flood = tf.math.reduce_max(predictions, axis=1)
+        return (predictions, max_flood)
+
+    def _update_temporal_inputs(
+        self, full_temp_input: tf.Tensor, t: int
+    ) -> tuple[tf.Tensor, tf.Tensor]:
+        """Updates temporal inputs for a new time step (prediction).
+
+        Returns the appropriate rainfall and flood map inputs for the t-th
+        prediction. The number n of flood maps and rainfall windows returned is
+        the minimum of self.params["n_flood_maps"] and t.
+
+        Args:
+            full_time_input: The [T, m] tensor of all rainfall windows. This
+                function retrieves the appropriate windows given the time step.
+            t: Time step in the autoregression.
+
+        Returns:
+            A tuple of tensors for the flood maps and rainfall, having shapes
+            [n, H, W, 1] and [n, m], respectively.
+        """
+        n = min(self.params["n_flood_maps"], t)
```

Comment:
Does it represent an error in the caller if they call this with a t > len(flood_maps) ?
I think it's generally better to raise errors than to try and code around bad inputs.

Reply:
This is to catch the case when we've made fewer than `n_flood_maps` predictions; in this case, we want to use all the flood maps available to us. Otherwise we would need to pad some zero flood maps at the beginning of this sequence.

Another way of looking at this is: The model looks at the most recent flood conditions to predict the flooding at the next time step. If we only have a few flood maps, we use all of them. If we have too many, we cap the number to the most recent `n_flood_maps` ones.